sys-boot/raspberrypi-firmware: bump to version 1.20190215

ConiKost opened the pull request:
Also overhauled eBuild and updated to EAPI=7.

Closes: https://bugs.gentoo.org/671238
Package-Manager: Portage-2.3.62, Repoman-2.3.12
Signed-off-by: Conrad Kostecki <conrad@kostecki.com>

ConiKost commented:
@mgorny: If you have some time, please have a look :)

juippis commented on sys-boot/raspberrypi-firmware/raspberrypi-firmware-1.20190215.ebuild:
```diff
@@ -0,0 +1,42 @@
+	insinto /boot
+	doins boot/*.{bin,dat,dtb,elf} "${FILESDIR}"/{cmdline,config}.txt
+
+	cat <<-EOF > "${T}"/99-raspberrypi-firmware || die
```
This isn't wrong and it's not a showstopper, but I'd like to make you aware that you don't have to use `${T}` in here. 
```
newenvd - "99${PN}.conf" <<- _EOF_
	CONFIG_PROTECT="/boot/config.txt /boot/cmdline.txt"
_EOF_
```

no need for die etc. Keep in mind for future :)

ConiKost replied:
Sounds good.

ConiKost commented:
@andrey-utkin Hey Andrey, here the PR, will add you to the metadata.xml too.

andrey-utkin commented:
Hi, thanks, will look in evening.

andrey-utkin commented:
1. I see config.txt has got replaced with a completely different one. What is the origin of your file? I will understand even if it's just your personal file. Preexisting one is "from archlinux arm sdcard image"which is also not terribly authoritative origin.

The Gentoo way is for package to fetch these files from some authoritative origin, rather than *being* an origin. But, to my dissatisfaction, I couldn't find any "reference" config.txt on raspberrypi.org or elsewhere, especially in a raw text file which we could put to SRC_URI.
I also see that at least config.txt is optional https://elinux.org/RPiconfig , so in principle we could decide to throw it away completely. Even though it is very useful to have such a file full of commented-out handy options.
The problem I see is that there's no good way to notice when the installed config becomes obsolete or suboptimal.

According to https://elinux.org/RPi_cmdline.txt it looks like cmdline.txt is also optional. So if we want, we can drop it, too. I cannot test this right now as I haven't deployed Gentoo for my RPi0W yet, but I would try that. 

-----------------

2. This isn't strictly required, nor it is a commonplace practice, but I find it much easier to make sense of changesets if ebuild cloning (version bumping) happens in one commit, and meaningful changes in other commits.

Don't worry about this for now - this is not a blocker, and I can just replay your changes in my preferred fashion if I want it so much.

-----------------

3. `eBuild` is not a conventional spelling.

-----------------

4. Just FYI - `:` doesn't need a semicolon `;` when it is on its own line, here:
```
src_compile() {
	:;
}
```

ConiKost commented:
> I see config.txt has got replaced with a completely different one. What is the origin of your file? I will understand even if it's just your personal file. Preexisting one is "from archlinux arm sdcard image"which is also not terribly authoritative origin.

There is no official config, as the repository doesn't contain one. The one from the standard image is also nicht official in that way ;-) Basically, the config is created by me, I took all available options and added them as an example, so a user can start directly. Since all options are commented out, nothing bad can happen.

> According to https://elinux.org/RPi_cmdline.txt it looks like cmdline.txt is also optional. So if we want, we can drop it, too. I cannot test this right now as I haven't deployed Gentoo for my RPi0W yet, but I would try that.

Currently, I've added the default, which is set on the raspberry pi upstream kernel. We could keep this empty, but IMHO it should be at least empty installed, so a user can start directly.

> This isn't strictly required, nor it is a commonplace practice, but I find it much easier to make sense of changesets if ebuild cloning (version bumping) happens in one commit, and meaningful changes in other commits.

Could you explain this? I don't see any commits, which should be squashed, based on that, what gentoo devs told me in the past.

> eBuild is not a conventional spelling.

You mean ```sys-boot/raspberrypi-firmware``` as name? I am open for changes, since I took this only, as it's the current name in portage.

> Just FYI - : doesn't need a semicolon ; when it is on its own line, here:

Thanks, changed.

andrey-utkin commented:
> There is no official config, as the repository doesn't contain one. The one from the standard image is also nicht official in that way ;-)

True, I see that.

> Basically, the config is created by me, I took all available options and added them as an example, so a user can start directly.

Please explain what you mean by "so a user can start directly".

> Since all options are commented out, nothing bad can happen.

Nothing bad happens to RPis users are booting, something slightly bad can happen to users' expectations.

**Accepting arbitrary changes (to configs in this case) is not sustainable.**

In this PR is that one arbitrary config.txt introduced by previous maintainers years ago is replaced by another arbitrary config.txt introduced by a new maintainer. So Gentoo users get different results and behaviour on an older installation and a new installation. This is no big deal usually, though. Also quite sometimes Gentoo maintainers author and maintain original configs to make some software play nice with the system (e.g. use logrotate etc.) if software doesn't ship a suitable config.

But just by having this original config sitting in the Gentoo repo, we are opening a possibility for future arbitrary changes. Say, if tomorrow a contributor proposes some new arbitrary changes to config.txt or indeed cmdline.txt, we should not require too mich justification to accept the changes, because there was not much justification in making earlier arbitrary changes to these files. And then later another contributor may propose opposite changes, giving his own justification :) Or another maintainer comes and replaces the txt files with his own personal ones again. See what I mean?

> it should be at least empty installed, so a user can start directly.

Again, I do not understand what exactly you mean by "so a user can start directly", please elaborate.

It is very important for me to understand your argument.

> > This isn't strictly required, nor it is a commonplace practice, but I find it much easier to make sense of changesets if ebuild cloning (version bumping) happens in one commit, and meaningful changes in other commits.
> 
> Could you explain this? I don't see any commits, which should be squashed, based on that, what gentoo devs told me in the past.

No I don't ask for squashing, quite the opposite!

For example, see how I usually do changes:

Clone the ebuild (lots of +/- lines, NO meaningful changes except for stable keywords dropping as needed to make repoman happy). 
https://github.com/gentoo/gentoo/commit/1281a108d0a381540c7ef25635db0cd27873ef87

Change SRC_URI. Meaningful change, little of stuff to read, easy to make sense immediately without additional tooling.
https://github.com/gentoo/gentoo/commit/60d3a3eaa176e4222d46b58e1844514fddfce6dc

Bump EAPI.
https://github.com/gentoo/gentoo/commit/95ee0e6715003dfbc26427f1cf9eb1e0d014075d

I'd like to back my position with opinion of Linus Torvalds:

https://lkml.org/lkml/2018/11/1/1312

> It would be _really_ nice to see that prep-phase be done in a way where each individual patch very obviously doesn't change semantics.

Prep-phase is ebuild cloning in this case.

The larger the total changeset, the more worried committers are about merging the changeset which doesn't meet this expectation. :)

> 
> > eBuild is not a conventional spelling.
> 
> You mean `sys-boot/raspberrypi-firmware` as name? I am open for changes, since I took this only, as it's the current name in portage.

I mean "eBuild" is not how the word "ebuild" is usually written.
"Ebuild" at the start of sentence, "ebuild" otherwise.

> > Just FYI - : doesn't need a semicolon ; when it is on its own line, here:
> 
> Thanks, changed.

I didn't mean you had to change it. While we are at it, any change to that place is worse than not changing it at all.

Just pointing out (I'm afraid my review already has too much bitterness!), that your changes to formatting of `src_compile()` and `src_configure()` are arbitrary and not needed. In every project using version control system it is advised to not commit such changes arbitrarily, even if you don't like the style you found.

The fact that cross-ebuild diffs are not directly visible when new ebuilds are created doesn't mean anybody should use this as opportunity to sneak in arbitrary changes.

andrey-utkin commented on sys-boot/raspberrypi-firmware/raspberrypi-firmware-1.20190215.ebuild:
```diff
@@ -0,0 +1,41 @@
+HOMEPAGE="https://github.com/raspberrypi/firmware"
+SRC_URI="https://github.com/raspberrypi/firmware/archive/${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="GPL-2 BSD"
```
This is not correct. Old LICENSE value was correct.
Everything but GPL stuff is exactly under a license known in Gentoo tree as raspberrypi-videocore-bin.

andrey-utkin commented on sys-boot/raspberrypi-firmware/raspberrypi-firmware-9999.ebuild:
```diff
@@ -1,55 +1,39 @@
+EGIT_REPO_URI="https://github.com/raspberrypi/firmware"
 
-LICENSE="GPL-2 raspberrypi-videocore-bin"
+LICENSE="GPL-2 BSD"
```
This is not correct. Old LICENSE value was correct.
Everything but GPL stuff is exactly under a license known in Gentoo tree as raspberrypi-videocore-bin.

ConiKost replied:
I think, this could be a question for the license team, as the ```raspberrypi-videocore-bin``` looks like a default BSD license, that's why I changed it in first place.

andrey-utkin commented:
I am currently selectively replaying parts of this PR. Will publish soon as another PR.

andrey-utkin commented:
Please see my take on this: https://github.com/gentoo/gentoo/pull/11465
I haven't tested this on hardware yet, will do this weekend.

andrey-utkin commented on sys-boot/raspberrypi-firmware/raspberrypi-firmware-1.20190215.ebuild:
```diff
@@ -0,0 +1,41 @@
+
+src_install() {
+	insinto /boot
+	doins boot/*.{bin,dat,dtb,elf} "${FILESDIR}"/{cmdline,config}.txt
```
Why did you drop installation of overlays dir?

ConiKost replied:
Because, when you compile a kernel, this will also install the overlays, that's why I added it instead to raspberrypi-image.

andrey-utkin commented:
Merged c1b8b3e9e39dbeab972bbf495dcba66e89807c75..8087386405b15949bf94973b0a4a81181173cbcd.

ConiKost commented:
> Please explain what you mean by "so a user can start directly".

A user don't have to copy that from the documentation, instead, can directly modify it to his purposes.

> Nothing bad happens to RPis users are booting, something slightly bad can happen to users' expectations.

Eh? Sorry, I don't see, what should change in terms of users expectations. Hey has now more options, but that's all.

> Accepting arbitrary changes (to configs in this case) is not sustainable.

> In this PR is that one arbitrary config.txt introduced by previous maintainers years ago is replaced by another arbitrary config.txt introduced by a new maintainer. So Gentoo users get different results and behaviour on an older installation and a new installation. This is no big deal usually, though. Also quite sometimes Gentoo maintainers author and maintain original configs to make some software play nice with the system (e.g. use logrotate etc.) if software doesn't ship a suitable config.

> But just by having this original config sitting in the Gentoo repo, we are opening a possibility for future arbitrary changes. Say, if tomorrow a contributor proposes some new arbitrary changes to config.txt or indeed cmdline.txt, we should not require too mich justification to accept the changes, because there was not much justification in making earlier arbitrary changes to these files. And then later another contributor may propose opposite changes, giving his own justification :) Or another maintainer comes and replaces the txt files with his own personal ones again. See what I mean?

Sorry, but I don't see your point. At least not for this case here. Since the config just got only completed, you can't talk here from arbitrary changes. I would accept that, If I would have just some new options and throuwn some of the old out of the config. That would be for me arbitrary.  I am not happy, that you discarded this completly! I don't think even, that those few options, which are currently set, are a good idea for the user experience, since not everyone want's to run a Pi as desktop.

> Again, I do not understand what exactly you mean by "so a user can start directly", please elaborate.
> It is very important for me to understand your argument.

Pretty simple. The user don't have to create the file, instead, he can directly insert his kernel parameters. That's what I meant this start directly. It's an official config file, which is used by the bootloader. So why not at least empty, if not the Raspberry Pi default?

> No I don't ask for squashing, quite the opposite!

Ok, I see, what you mean. That's interesseting, as every other gentoo dev wanted me, to do my change in logical commits, not every commit for a change, as you did in your PRs. Basically, as I did it so far. If you want, I can do this in future for you, no problem for me.

> I mean "eBuild" is not how the word "ebuild" is usually written.
> "Ebuild" at the start of sentence, "ebuild" otherwise.

Just a bad behaviour by me ;-) I sometimes forget this.

> Just pointing out (I'm afraid my review already has too much bitterness!), that your changes to formatting of src_compile() and src_configure() are arbitrary and not needed. In every project using version control system it is advised to not commit such changes arbitrarily, even if you don't like the style you found.

Sorry, but I disagree here. When I am supposed to be the maintainer, I've my style, how I like to read it, since I changing it. I don't consider this arbitrary. I would accept this, when I am not the maintainer. Yes, I know, you don't see is this way.

andrey-utkin commented:
On Sat, Mar 30, 2019 at 04:48:52AM -0700, Conrad Kostecki wrote:
> Because, when you compile a kernel, this will also install the overlays, that's why I added it instead to raspberrypi-image.

I wasn't aware of this issue.
You haven't explained this anywhere in your commit messages.
I am happy to accept such change, since I know now.

andrey-utkin commented:
On Sat, Mar 30, 2019 at 05:04:52AM -0700, Conrad Kostecki wrote:
> > Please explain what you mean by "so a user can start directly".
> 
> A user don't have to copy that from the documentation, instead, can directly modify it to his purposes.

I see now.

> > Nothing bad happens to RPis users are booting, something slightly bad can happen to users' expectations.
> 
> Eh? Sorry, I don't see, what should change in terms of users expectations. Hey has now more options, but that's all.

well, `diff` shown me that your config files are completely different
from the old ones. Since I haven't had any specific explanation what's
going on (such things normally go in commit messages BTW), I assumed
that they ARE different. From this I concluded that users who would have
merged your new version would potentially get different behaviour to
what they got, say, a day before when they merged an old ebuild. Maybe
not a big deal really, but I wanted to avoid this, especially the value
of the change hasn't been explained.

> > Accepting arbitrary changes (to configs in this case) is not sustainable.
> 
> > In this PR is that one arbitrary config.txt introduced by previous maintainers years ago is replaced by another arbitrary config.txt introduced by a new maintainer. So Gentoo users get different results and behaviour on an older installation and a new installation. This is no big deal usually, though. Also quite sometimes Gentoo maintainers author and maintain original configs to make some software play nice with the system (e.g. use logrotate etc.) if software doesn't ship a suitable config.
> 
> > But just by having this original config sitting in the Gentoo repo, we are opening a possibility for future arbitrary changes. Say, if tomorrow a contributor proposes some new arbitrary changes to config.txt or indeed cmdline.txt, we should not require too mich justification to accept the changes, because there was not much justification in making earlier arbitrary changes to these files. And then later another contributor may propose opposite changes, giving his own justification :) Or another maintainer comes and replaces the txt files with his own personal ones again. See what I mean?
> 
> Sorry, but I don't see your point. At least not for this case here. Since the config just got only completed, you can't talk here from arbitrary changes. I would accept that, If I would have just some new options and throuwn some of the old out of the config. That would be for me arbitrary.

But you have done exactly that, haven't you?

Let me double check.

The old config boils down to:

```
gpu_mem_512=316
gpu_mem_256=128
cma_lwm=16
cma_hwm=32
cma_offline_start=16
```

The new config doesn't define any parameter values, it only lists all
options commented out.

Maybe all of the settings present in the old config are bogus as these
are default values? Let's check with your config.

```
$ grep 'gpu_mem_512' config.txt
# gpu_mem_512=448
$ grep 'gpu_mem_256' config.txt
# gpu_mem_256=192
$ grep 'cma_lwm' config.txt
$ grep 'cma_hwm' config.txt
$ grep 'cma_offline_start' config.txt
```

Two options will have different values.
Three options explicitly set before are not set, and they are not listed
in your config at all.

> I am not happy, that you discarded this completly! I don't think even, that those few options, which are currently set, are a good idea for the user experience, since not everyone want's to run a Pi as desktop.

I am not a fan of the old config, too. To me, it seems too...
*arbitrary*.

> Pretty simple. The user don't have to create the file, instead, he can directly insert his kernel parameters. That's what I meant this start directly. It's an official config file, which is used by the bootloader. So why not at least empty, if not the Raspberry Pi default?

I would be totally happy with shipping a real official default config.
Where do I get it?

> > No I don't ask for squashing, quite the opposite!
> 
> Ok, I see, what you mean. That's interesseting, as every other gentoo dev wanted me, to do my change in logical commits, not every commit for a change, as you did in your PRs.

I didn't get the difference you meant to highlight here.

> Basically, as I did it so far. If you want, I can do this in future for you, no problem for me.

Yes, I think it's worth trying, and I even think other Gentoo devs,
including proxy maints, will be happy too with commits which really
address one little thing a time and explain well what was the motivation
and how the change helps.

> > Just pointing out (I'm afraid my review already has too much bitterness!), that your changes to formatting of src_compile() and src_configure() are arbitrary and not needed. In every project using version control system it is advised to not commit such changes arbitrarily, even if you don't like the style you found.
> 
> Sorry, but I disagree here. When I am supposed to be the maintainer, I've my style, how I like to read it, since I changing it. I don't consider this arbitrary. I would accept this, when I am not the maintainer. Yes, I know, you don't see is this way.

Gentoo Development Guide says:

> Unnecessarily changing the syntax of an ebuild can cause complications for others. Syntax changes should only be done if there is a real benefit, such as faster compilation, improved information for the end user, or compliance with Gentoo policies.

andrey-utkin commented:
On Sat, Mar 30, 2019 at 04:51:09AM -0700, Conrad Kostecki wrote:
> ConiKost commented on this pull request.
> >  
> -LICENSE="GPL-2 raspberrypi-videocore-bin"
> +LICENSE="GPL-2 BSD"
> 
> I think, this could be a question for the license team, as the ```raspberrypi-videocore-bin``` looks like a default BSD license, that's why I changed it in first place.

No, it's not the same as BSD, it is quite different. This is very
restrictive license in fact:

> This software may only be used for the purposes of developing for, 
> running or using a Raspberry Pi device.

ConiKost commented:
> I wasn't aware of this issue. You haven't explained this anywhere in your commit messages. I am happy to accept such change, since I know now.

Well, this was in my PR, you could ask :P Anyway, thanks.

> well, `diff` shown me that your config files are completely different from the old ones. Since I haven't had any specific explanation what's going on (such things normally go in commit messages BTW), I assumed that they ARE different. From this I concluded that users who would have merged your new version would potentially get different behaviour to what they got, say, a day before when they merged an old ebuild. Maybe not a big deal really, but I wanted to avoid this, especially the value of the change hasn't been explained.

Well, I asumued, it was obvioius those changes :) Will try to explain, but later on.

> But you have done exactly that, haven't you?

Yes and no. Those definitions about memory are just pointless. With setting them in first place means, we are forcing a user in a specific scenario of the Pi. If you want to run a Pi in "server mode", you don't want to set. So its much better do use the Pis default and let a user to decide, what he want's to do.

And the missing parameters aren't anymore documented in the official documentation, which means for me, they got dropped.

> I would be totally happy with shipping a real official default config. Where do I get it?
Depends, what you consider default. For me, I took the options, which are coded for the upstream Pi kernel, that's what I consider official. But you won't find the file in the git repos, as it does not exist there.

> I didn't get the difference you meant to highlight here.
You are splitting nearly every single change in a commit. Other devs asks me not to do this, just jew commits. Hope you see, what I meant now :)

>  Unnecessarily changing the syntax of an ebuild can cause complications for others. Syntax changes should only be done if there is a real benefit, such as faster compilation, improved information for the end user, or compliance with Gentoo policies.

Simple: It's better readable for me.

> No, it's not the same as BSD, it is quite different.

I see, thanks.

andrey-utkin commented:
On Sun, Mar 31, 2019 at 03:56:16PM -0700, Conrad Kostecki wrote:
> > I wasn't aware of this issue. You haven't explained this anywhere in your commit messages. I am happy to accept such change, since I know now.
> 
> Well, this was in my PR, you could ask :P Anyway, thanks.

Reviews should be straightforward and authors should really work hard to
make it so.

Discussions may arise because of some tricky issues. Reviewing should
not involve guessing author's intention about core topic of the change.

> > But you have done exactly that, haven't you?
> 
> Yes and no. Those definitions about memory are just pointless. With setting them in first place means, we are forcing a user in a specific scenario of the Pi. If you want to run a Pi in "server mode", you don't want to set. So its much better do use the Pis default and let a user to decide, what he want's to do.
> 
> And the missing parameters aren't anymore documented in the official documentation, which means for me, they got dropped.

Have they been documented before? Guess could be they never were.

> > I would be totally happy with shipping a real official default config. Where do I get it?
> Depends, what you consider default. For me, I took the options, which are coded for the upstream Pi kernel, that's what I consider official. But you won't find the file in the git repos, as it does not exist there.

If you made it, and not on behalf of RPi Foundation, it's not official.

> >  Unnecessarily changing the syntax of an ebuild can cause complications for others. Syntax changes should only be done if there is a real benefit, such as faster compilation, improved information for the end user, or compliance with Gentoo policies.
> 
> Simple: It's better readable for me.

Well, Gentoo is a high-scale collaborative project and it's not all
about you.

Conrad, I'm ashamed of myself for all of this boring mentorship.
Part of your dissatisfation can definitely be attributed to my
positional thinking "I have commit access and you don't". Become Gentoo
Developer already please-please-please, and you will be free from this!
Both of us will save a lot of time! :)

ConiKost commented:
> Discussions may arise because of some tricky issues. Reviewing should not involve guessing author's intention about core topic of the change.

Sorry, I thought, you were aware of this :D

> Have they been documented before? Guess could be they never were.

I can't prove, since there is no history, but I can, hardly, remember, that those cma_* were once documented there.

> If you made it, and not on behalf of RPi Foundation, it's not official.

Eh? The official Pi distribution ships that. But as I said: I am also happy, if we only installed a empty file.

> Well, Gentoo is a high-scale collaborative project and it's not all about you.

Sure, but on the other hand, when I am supposed to take maintenance ownership on a package, I would still see this something, which makes it more easier for me :)

> Conrad, I'm ashamed of myself for all of this boring mentorship. Part of your dissatisfation can definitely be attributed to my positional thinking "I have commit access and you don't". Become Gentoo Developer already please-please-please, and you will be free from this! Both of us will save a lot of time! :)

Yeah, I know, I am late on it, have to finish finally my quizzes, still on my todo list. But don't get me wrong: I am not angry or something on you, we just have a discussion here :)

andrey-utkin commented:
On Mon, Apr 01, 2019 at 10:08:09AM +0000, Conrad Kostecki wrote:
> > Have they been documented before? Guess could be they never were.
> 
> I can't prove, since there is no history, but I can, hardly, remember, that those cma_* were once documented there.

OK. But obviously I am not so familiar with this and it was a red flag
for me.

> > If you made it, and not on behalf of RPi Foundation, it's not official.
> 
> Eh? The official Pi distribution ships that. But as I said: I am also happy, if we only installed a empty file.

This is a credibility issue. One way to solve it is to put what they
ship to SRC_URI and install that.

Also happy to install an empty file, or, as I initially proposed, to
install nothing.

ConiKost commented:
> This is a credibility issue. One way to solve it is to put what they ship to SRC_URI and install that.

None, since its per user.

> Also happy to install an empty file, or, as I initially proposed, to install nothing.

Not installing would be the worst experience for a user. Let's keep installing an empty file.

andrey-utkin commented:
@ConiKost I should have mentioned in our config.txt argument that it would be very Gentoo-Way and very good thing to do in general if you submitted your config.txt in a pull request for the upstream github repo, so that your config could *become* official in strict sense of this word.
